0.68.0: go panic w/ multiple bases/layers

continuation of #615

Running same exact scenario as described here: 

https://github.com/roboll/helmfile/issues/615#issuecomment-496694205

```
bash-4.4$ ./helmfile -version
helmfile version v0.68.0

bash-4.4$ ./helmfile --log-level debug -f deployments.yaml -n default -e test1 apply

processing file "deployments.yaml" in directory "."
first-pass rendering starting for "deployments.yaml.part.0": inherited=&{test1 map[]}, overrode=<nil>
first-pass uses: &{test1 map[]}
error in first-pass rendering: result of "deployments.yaml.part.0":
 0: bases:
 1:   - envs.yaml
 2:

first-pass produced: &{test1 map[]}
first-pass rendering result of "deployments.yaml.part.0": {test1 map[]}
second-pass rendering result of "deployments.yaml.part.0":
 0: bases:
 1:   - envs.yaml
 2:

first-pass rendering starting for "envs.yaml.part.0": inherited=&{test1 map[]}, overrode=<nil>
first-pass uses: &{test1 map[]}
first-pass produced: &{test1 map[]}
panic: json: unsupported type: map[interface {}]interface {} [recovered]
	panic: json: unsupported type: map[interface {}]interface {}

goroutine 1 [running]:
github.com/urfave/cli.HandleAction.func1(0xc000240de0)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20160620154522-6011f165dc28/app.go:474 +0x287
panic(0x14583a0, 0xc000061d30)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/roboll/helmfile/environment.Environment.DeepCopy(0x7ffeefbffa27, 0x5, 0xc000085c50, 0x13199d1, 0xc000148000, 0x0)
	/home/circleci/workspace/helmfile/environment/environment.go:18 +0x12b
github.com/roboll/helmfile/environment.(*Environment).Merge(0xc00022a6a8, 0x0, 0x14d9408, 0x17, 0xc00023f7e8)
	/home/circleci/workspace/helmfile/environment/environment.go:38 +0x142
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).twoPassRenderTemplateToYaml(0xc000150000, 0xc0001ac0c0, 0x0, 0x14d1ccc, 0x1, 0xc0001ce000, 0x10, 0xc0001cc000, 0x66, 0x266, ...)
	/home/circleci/workspace/helmfile/pkg/app/two_pass_renderer.go:86 +0x133
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).renderTemplatesToYamlWithEnv(...)
	/home/circleci/workspace/helmfile/pkg/app/two_pass_renderer.go:62
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).renderTemplatesToYaml(0xc000150000, 0x14d1ccc, 0x1, 0xc0001ce000, 0x10, 0xc0001cc000, 0x66, 0x266, 0xc0001ac0a0, 0x1, ...)
	/home/circleci/workspace/helmfile/pkg/app/two_pass_renderer.go:58 +0xd2
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).renderAndLoad(0xc000150000, 0x0, 0x0, 0x14d1ccc, 0x1, 0xc0000250f0, 0x9, 0xc0001cc000, 0x66, 0x266, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:170 +0x6d3
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).loadFileWithOverrides(0xc000150000, 0x0, 0x0, 0x14d1ccc, 0x1, 0xc000024f80, 0x9, 0x0, 0x1974bd8, 0x1, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:110 +0x1f4
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).loadFile(...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:90
github.com/roboll/helmfile/state.(*StateCreator).loadBases(0xc000031680, 0x0, 0xc0001181c0, 0x14d1ccc, 0x1, 0x1, 0xc000152000, 0x10)
	/home/circleci/workspace/helmfile/state/create.go:146 +0xe2
github.com/roboll/helmfile/state.(*StateCreator).ParseAndLoad(0xc000031680, 0xc000142100, 0x15, 0x40, 0x14d1ccc, 0x1, 0xc000152000, 0x10, 0x7ffeefbffa27, 0x5, ...)
	/home/circleci/workspace/helmfile/state/create.go:135 +0xe6
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).load(0xc000150000, 0xc000142100, 0x15, 0x40, 0x14d1ccc, 0x1, 0xc000152000, 0x10, 0xc000158501, 0x0, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:144 +0x11a
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).renderAndLoad(0xc000150000, 0x0, 0x0, 0x14d1ccc, 0x1, 0xc000152000, 0x10, 0xc000154000, 0x154, 0x354, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:181 +0x345
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).loadFileWithOverrides(0xc000150000, 0x0, 0x0, 0x14d1ccc, 0x1, 0x7ffeefbffa08, 0x10, 0x1890101, 0x1977848, 0x203000, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:110 +0x1f4
github.com/roboll/helmfile/pkg/app.(*desiredStateLoader).Load(0xc000150000, 0x7ffeefbffa08, 0x10, 0x18a6328, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/circleci/workspace/helmfile/pkg/app/desired_state_file_loader.go:59 +0x473
github.com/roboll/helmfile/pkg/app.(*App).loadDesiredStateFromYaml(0xc00001f7c0, 0x7ffeefbffa08, 0x10, 0xc0000cc4a8, 0x1, 0x1, 0xc0000cc320, 0xc000000000, 0xc0000cc378)
	/home/circleci/workspace/helmfile/pkg/app/app.go:140 +0x1c6
github.com/roboll/helmfile/pkg/app.(*App).VisitDesiredStates.func1(0x7ffeefbffa08, 0x10, 0xc00014e080, 0x7a, 0x0, 0x0)
	/home/circleci/workspace/helmfile/pkg/app/app.go:147 +0xf3
github.com/roboll/helmfile/pkg/app.(*App).visitStateFiles.func1(0x0, 0x0)
	/home/circleci/workspace/helmfile/pkg/app/app.go:111 +0x9a
github.com/roboll/helmfile/pkg/app.(*App).within(0xc00001f7c0, 0x14d1ccc, 0x1, 0xc000031640, 0xc0000cc820, 0x2)
	/home/circleci/workspace/helmfile/pkg/app/app.go:53 +0x722
github.com/roboll/helmfile/pkg/app.(*App).visitStateFiles(0xc00001f7c0, 0x7ffeefbffa08, 0x10, 0xc00006c720, 0x10, 0x1475180)
	/home/circleci/workspace/helmfile/pkg/app/app.go:105 +0x29f
github.com/roboll/helmfile/pkg/app.(*App).VisitDesiredStates(0xc00001f7c0, 0x7ffeefbffa08, 0x10, 0x18a6328, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/home/circleci/workspace/helmfile/pkg/app/app.go:146 +0xd4
github.com/roboll/helmfile/pkg/app.(*App).VisitDesiredStatesWithReleasesFiltered(0xc00001f7c0, 0x7ffeefbffa08, 0x10, 0xc00000d4a0, 0x10, 0x0)
	/home/circleci/workspace/helmfile/pkg/app/app.go:235 +0x10e
github.com/roboll/helmfile/cmd.FindAndIterateOverDesiredStatesUsingFlagsWithReverse(0xc0000bef00, 0xc00000d400, 0xc00000d480, 0x0, 0xc000085860)
	/home/circleci/workspace/helmfile/cmd/cmd.go:77 +0xe6
main.findAndIterateOverDesiredStatesUsingFlags(...)
	/home/circleci/workspace/helmfile/main.go:633
main.main.func7(0xc0000bef00, 0x0, 0x0)
	/home/circleci/workspace/helmfile/main.go:356 +0x8d
reflect.Value.call(0x14464e0, 0x14efde0, 0x13, 0x14d21a8, 0x4, 0xc0000ccd80, 0x1, 0x1, 0xc000030080, 0x1011bc3, ...)
	/usr/local/go/src/reflect/value.go:447 +0x461
reflect.Value.Call(0x14464e0, 0x14efde0, 0x13, 0xc0000ccd80, 0x1, 0x1, 0xc00008d040, 0xc00008d098, 0x140)
	/usr/local/go/src/reflect/value.go:308 +0xa4
github.com/urfave/cli.HandleAction(0x14464e0, 0x14efde0, 0xc0000bef00, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20160620154522-6011f165dc28/app.go:483 +0x1ff
github.com/urfave/cli.Command.Run(0x14d2666, 0x5, 0x0, 0x0, 0x0, 0x0, 0x0, 0x14e85ba, 0x3f, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20160620154522-6011f165dc28/command.go:186 +0x8d1
github.com/urfave/cli.(*App).Run(0xc000001200, 0xc00001e0a0, 0xa, 0xa, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v0.0.0-20160620154522-6011f165dc28/app.go:237 +0x601
main.main()
	/home/circleci/workspace/helmfile/main.go:592 +0x21e6
```

--- a/environment/environment.go
+++ b/environment/environment.go
@@ -2,6 +2,7 @@ package environment
 
 import (
 	"github.com/imdario/mergo"
+	"github.com/roboll/helmfile/pkg/maputil"
 	"gopkg.in/yaml.v2"
 )
 
@@ -19,6 +20,10 @@ func (e Environment) DeepCopy() Environment {
 	}
 	var values map[string]interface{}
 	if err := yaml.Unmarshal(bytes, &values); err != nil {
+		panic(err)
+	}
+	values, err = maputil.CastKeysToStrings(values)
+	if err != nil {
 		panic(err)
 	}
 	return Environment{

--- a/environment/environment.go
+++ b/environment/environment.go
@@ -1,8 +1,8 @@
 package environment
 
 import (
-	"encoding/json"
 	"github.com/imdario/mergo"
+	"gopkg.in/yaml.v2"
 )
 
 type Environment struct {
@@ -13,12 +13,12 @@ type Environment struct {
 var EmptyEnvironment Environment
 
 func (e Environment) DeepCopy() Environment {
-	bytes, err := json.Marshal(e.Values)
+	bytes, err := yaml.Marshal(e.Values)
 	if err != nil {
 		panic(err)
 	}
 	var values map[string]interface{}
-	if err := json.Unmarshal(bytes, &values); err != nil {
+	if err := yaml.Unmarshal(bytes, &values); err != nil {
 		panic(err)
 	}
 	return Environment{

--- a/pkg/app/app_test.go
+++ b/pkg/app/app_test.go
@@ -1417,3 +1417,77 @@ releases:
 		}
 	}
 }
+
+func TestLoadDesiredStateFromYaml_MultiPartTemplate_SprigDictFuncs(t *testing.T) {
+	type testcase struct {
+		state    string
+		expr     string
+		expected string
+	}
+	stateInline := `
+environments:
+  default:
+    values:
+    - foo: FOO
+      bar: { "baz": "BAZ" }
+---
+releases:
+- name: %s
+  chart: stable/nginx
+`
+	stateExternal := `
+environments:
+  default:
+    values:
+    - 1.yaml
+    - 2.yaml
+---
+releases:
+- name: %s
+  chart: stable/nginx
+`
+	testcases := []testcase{
+		{stateInline, `{{ getOrNil "foo" .Environment.Values }}`, `FOO`},
+		{stateInline, `{{ getOrNil "baz" (getOrNil "bar" .Environment.Values) }}`, `BAZ`},
+		{stateInline, `{{ if hasKey .Environment.Values "foo" }}{{ .Environment.Values.foo }}{{ end }}`, `FOO`},
+		{stateInline, `{{ if hasKey .Environment.Values "bar" }}{{ .Environment.Values.bar.baz }}{{ end }}`, `BAZ`},
+		{stateInline, `{{ if (keys .Environment.Values | has "foo") }}{{ .Environment.Values.foo }}{{ end }}`, `FOO`},
+		// See https://github.com/roboll/helmfile/issues/624
+		// This fails when .Environment.Values.bar is not map[string]interface{}. At the time of #624 it was map[interface{}]interface{}, which sprig's dict funcs don't support.
+		{stateInline, `{{ if (keys .Environment.Values | has "bar") }}{{ if (keys .Environment.Values.bar | has "baz") }}{{ .Environment.Values.bar.baz }}{{ end }}{{ end }}`, `BAZ`},
+		{stateExternal, `{{ getOrNil "foo" .Environment.Values }}`, `FOO`},
+		{stateExternal, `{{ getOrNil "baz" (getOrNil "bar" .Environment.Values) }}`, `BAZ`},
+		{stateExternal, `{{ if hasKey .Environment.Values "foo" }}{{ .Environment.Values.foo }}{{ end }}`, `FOO`},
+		{stateExternal, `{{ if hasKey .Environment.Values "bar" }}{{ .Environment.Values.bar.baz }}{{ end }}`, `BAZ`},
+		{stateExternal, `{{ if (keys .Environment.Values | has "foo") }}{{ .Environment.Values.foo }}{{ end }}`, `FOO`},
+		// See https://github.com/roboll/helmfile/issues/624
+		{stateExternal, `{{ if (keys .Environment.Values | has "bar") }}{{ if (keys .Environment.Values.bar | has "baz") }}{{ .Environment.Values.bar.baz }}{{ end }}{{ end }}`, `BAZ`},
+	}
+	for i := range testcases {
+		tc := testcases[i]
+		statePath := "/path/to/helmfile.yaml"
+		stateContent := fmt.Sprintf(tc.state, tc.expr)
+		testFs := state.NewTestFs(map[string]string{
+			statePath:         stateContent,
+			"/path/to/1.yaml": `foo: FOO`,
+			"/path/to/2.yaml": `bar: { "baz": "BAZ" }`,
+		})
+		app := &App{
+			readFile: testFs.ReadFile,
+			glob:     testFs.Glob,
+			abs:      testFs.Abs,
+			Env:      "default",
+			Logger:   helmexec.NewLogger(os.Stderr, "debug"),
+			Reverse:  true,
+		}
+		st, err := app.loadDesiredStateFromYaml(statePath)
+
+		if err != nil {
+			t.Fatalf("unexpected error at %d: %v", i, err)
+		}
+
+		if st.Releases[0].Name != tc.expected {
+			t.Errorf("unexpected releases[0].name at %d: expected=%s, got=%s", i, tc.expected, st.Releases[0].Name)
+		}
+	}
+}

--- a/pkg/app/app_test.go
+++ b/pkg/app/app_test.go
@@ -1356,3 +1356,64 @@ releases:
 		t.Errorf("unexpected releases[0].chart: expected=BAR, got=%s", st.Releases[0].Chart)
 	}
 }
+
+// See https://github.com/roboll/helmfile/issues/623
+func TestLoadDesiredStateFromYaml_MultiPartTemplate_MergeMapsVariousKeys(t *testing.T) {
+	type testcase struct {
+		overrideValues interface{}
+		expected       string
+	}
+	testcases := []testcase{
+		{map[interface{}]interface{}{"foo": "FOO"}, `FOO`},
+		{map[interface{}]interface{}{"foo": map[interface{}]interface{}{"foo": "FOO"}}, `map[foo:FOO]`},
+		{map[interface{}]interface{}{"foo": map[string]interface{}{"foo": "FOO"}}, `map[foo:FOO]`},
+		{map[interface{}]interface{}{"foo": []interface{}{"foo"}}, `[foo]`},
+		{map[interface{}]interface{}{"foo": "FOO"}, `FOO`},
+	}
+	for i := range testcases {
+		tc := testcases[i]
+		statePath := "/path/to/helmfile.yaml"
+		stateContent := `
+environments:
+  default:
+    values:
+    - 1.yaml
+    - 2.yaml
+---
+releases:
+- name: {{ .Environment.Values.foo | quote }}
+  chart: {{ .Environment.Values.bar | quote }}
+`
+		testFs := state.NewTestFs(map[string]string{
+			statePath:         stateContent,
+			"/path/to/1.yaml": `bar: ["bar"]`,
+			"/path/to/2.yaml": `bar: ["BAR"]`,
+		})
+		app := &App{
+			readFile: testFs.ReadFile,
+			glob:     testFs.Glob,
+			abs:      testFs.Abs,
+			Env:      "default",
+			Logger:   helmexec.NewLogger(os.Stderr, "debug"),
+			Reverse:  true,
+		}
+		opts := LoadOpts{
+			CalleePath: statePath,
+			Environment: state.SubhelmfileEnvironmentSpec{
+				OverrideValues: []interface{}{tc.overrideValues},
+			},
+		}
+		st, err := app.loadDesiredStateFromYaml(statePath, opts)
+
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+
+		if st.Releases[0].Name != tc.expected {
+			t.Errorf("unexpected releases[0].name: expected=%s, got=%s", tc.expected, st.Releases[0].Name)
+		}
+		if st.Releases[0].Chart != "[BAR]" {
+			t.Errorf("unexpected releases[0].chart: expected=BAR, got=%s", st.Releases[0].Chart)
+		}
+	}
+}

--- a/pkg/app/two_pass_renderer.go
+++ b/pkg/app/two_pass_renderer.go
@@ -80,7 +80,7 @@ func (r *desiredStateLoader) twoPassRenderTemplateToYaml(inherited, overrode *en
 	renderedEnv := r.renderEnvironment(initEnv, baseDir, filename, content)
 
 	if r.logger != nil {
-		r.logger.Debugf("first-pass produced: %v", initEnv)
+		r.logger.Debugf("first-pass produced: %v", renderedEnv)
 	}
 
 	finalEnv, err := renderedEnv.Merge(overrode)

--- a/pkg/maputil/maputil.go
+++ b/pkg/maputil/maputil.go
@@ -1,0 +1,50 @@
+package maputil
+
+import "fmt"
+
+func CastKeysToStrings(s interface{}) (map[string]interface{}, error) {
+	new := map[string]interface{}{}
+	switch src := s.(type) {
+	case map[interface{}]interface{}:
+		for k, v := range src {
+			var str_k string
+			switch typed_k := k.(type) {
+			case string:
+				str_k = typed_k
+			default:
+				return nil, fmt.Errorf("unexpected type of key in map: expected string, got %T: value=%v, map=%v", typed_k, typed_k, src)
+			}
+
+			var casted_v interface{}
+			switch typed_v := v.(type) {
+			case map[interface{}]interface{}:
+				tmp, err := CastKeysToStrings(typed_v)
+				if err != nil {
+					return nil, err
+				}
+				casted_v = tmp
+			default:
+				casted_v = typed_v
+			}
+
+			new[str_k] = casted_v
+		}
+	case map[string]interface{}:
+		for k, v := range src {
+			var casted_v interface{}
+			switch typed_v := v.(type) {
+			case map[interface{}]interface{}:
+				tmp, err := CastKeysToStrings(typed_v)
+				if err != nil {
+					return nil, err
+				}
+				casted_v = tmp
+			default:
+				casted_v = typed_v
+			}
+
+			new[k] = casted_v
+		}
+	}
+	return new, nil
+}

--- a/state/environment_values_loader.go
+++ b/state/environment_values_loader.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"github.com/imdario/mergo"
 	"github.com/roboll/helmfile/environment"
+	"github.com/roboll/helmfile/pkg/maputil"
 	"github.com/roboll/helmfile/tmpl"
 	"gopkg.in/yaml.v2"
 	"path/filepath"
@@ -53,14 +54,9 @@ func (ld *EnvironmentValuesLoader) LoadEnvironmentValues(missingFileHandler *str
 				}
 			}
 		case map[interface{}]interface{}:
-			m := map[string]interface{}{}
-			for k, v := range typedValue {
-				switch typedKey := k.(type) {
-				case string:
-					m[typedKey] = v
-				default:
-					return nil, fmt.Errorf("unexpected type of key in inline environment values %v: expected string, got %T", typedValue, typedKey)
-				}
+			m, err := maputil.CastKeysToStrings(typedValue)
+			if err != nil {
+				return nil, err
 			}
 			if err := mergo.Merge(&envVals, &m, mergo.WithOverride); err != nil {
 				return nil, fmt.Errorf("failed to merge %v: %v", typedValue, err)

--- a/state/environment_values_loader.go
+++ b/state/environment_values_loader.go
@@ -67,7 +67,7 @@ func (ld *EnvironmentValuesLoader) LoadEnvironmentValues(missingFileHandler *str
 			}
 			continue
 		default:
-			return nil, fmt.Errorf("unexpected type of values entry: %T", typedValue)
+			return nil, fmt.Errorf("unexpected type of value: value=%v, type=%T", typedValue, typedValue)
 		}
 	}
 

--- a/state/state.go
+++ b/state/state.go
@@ -1319,7 +1319,7 @@ func (st *HelmState) generateTemporaryValuesFiles(values []interface{}, missingF
 			}
 			generatedFiles = append(generatedFiles, valfile.Name())
 		default:
-			return nil, fmt.Errorf("unexpected type of values entry: %T", typedValue)
+			return nil, fmt.Errorf("unexpected type of value: value=%v, type=%T", typedValue, typedValue)
 		}
 	}
 	return generatedFiles, nil

